driver-core/src/main/java/com/datastax/driver/core/NettyOptions.java

```diff
@@ -0,0 +1,194 @@
+ * its methods.
+ * <p>
+ * Typically, clients would register this class with {@link Cluster#builder()}:
+ *
+ * <pre>
+ *     NettyOptions nettyOptions = ...
+ *     Cluster cluster = Cluster.builder()
+ *          .addContactPoint(...)
+ *          .withNettyOptions(nettyOptions)
+ *          .build();
+ * </pre>
+ *
+ * <strong>Extending the NettyOptions API</strong>
+ * <p>
+ * Contrary to other driver options, the options available in this class should
+ * be considered as advanced features and as such, <em>they should only be
+ * modified by expert users</em>.
+ * <p>
+ * <strong>A misconfiguration introduced by the means of this API can have unexpected results
+ * and cause the driver to completely fail to connect.</strong>
+ * <p>
+ * Moreover, since versions 2.0.9 and 2.1.4 (see JAVA-538),
+ * the driver is available in two different flavors: with a standard Maven dependency on Netty,
+ * or with a "shaded" (internalized) Netty dependency.
+ * <p>
+ * Given that NettyOptions API exposes Netty classes ({@link SocketChannel}, etc.),
+ * <em>it should only be extended by clients using the non-shaded
+ * version of driver</em>.
+ * <p>
+ * <strong>Extending this API with shaded Netty classes is not supported,
+ * and in particular for OSGi applications, it is likely that such a configuration would lead to
+ * compile and/or runtime errors.</strong>
+ *
+ * @jira_ticket JAVA-640
+ * @since 2.0.10
+ */
+public class NettyOptions {
+
+    /**
+     * The default instance of {@link NettyOptions} to use.
+     */
+    public static final NettyOptions DEFAULT_INSTANCE = new NettyOptions();
+
+    /**
+     * Return the {@link EventLoopGroup} instance to use.
+     * <p>
+     * This hook is invoked only once at {@link Cluster} initialization;
+     * the returned instance will be kept in use throughout the cluster lifecycle.
+     * <p>
+     * Typically, implementors would return a newly-created instance;
+     * it is however possible to re-use a shared instance, but in this
+     * case implementors should also override {@link #onClusterClose(EventLoopGroup)}
+     * to prevent the shared instance to be closed when the cluster is closed.
+     * <p>
+     * The default implementation returns a new instance of {@link io.netty.channel.epoll.EpollEventLoopGroup}
+     * if {@link NettyUtil#isEpollAvailable() epoll is available},
+     * or {@link NioEventLoopGroup} otherwise.
+     *
+     * @param threadFactory The {@link ThreadFactory} to use when creating a new {@link EventLoopGroup} instance;
+     *                      The driver will provide its own internal thread factory here.
+     *                      It is safe to ignore it and use another thread factory.
+     * @return the {@link EventLoopGroup} instance to use.
+     */
+    public EventLoopGroup eventLoopGroup(ThreadFactory threadFactory) {
+        return NettyUtil.newEventLoopGroupInstance(threadFactory);
```

Comment:
@tolbertam you were actually right: we cannot import anything from native-epoll here because if users explicitly exclude it we would have runtime errors here. So we do need reflection after all, but I moved this low-level code to NettyUtil.